docs/source/jit_language_reference_v2.rst

```diff
@@ -29,6 +29,213 @@
+
+Local variables
+"""""""""""""""
+Local variables can be annotated according to Python3 typing module annotation rule, i.e.,
+
+::
+
+    LocalVarAnnotation := Identifier [":" TSType] "=" Expr
+
+In general, types of local variables can be automatically inferred. In some cases, however, programmers may need to annotate a multi-type for local variables
+that may be associated with different concrete types. Typical multi-types include ``Optional[T]`` and ``Any``.
+
+**Example**
+
+::
+
+    import torch
+
+    def f(a, setVal: bool):
+        value: Optional[torch.Tensor] = None
+        if setVal:
+            value = a
+        return value
+
+    ones = torch.ones([6])
+    m = torch.jit.script(f)
+    print("TorchScript:", m(ones, True), m(ones, False))
+
+Instance data attributes
+""""""""""""""""""""""""
+For ``ModuleType`` classes, instance data attributes can be annotated according to Python3 typing module annotation rules. Instance data attributes can be annotated (optionally) as final
+via ``torch.jit.Final``.
+
+::
+
+    "class" ClassIdentifier "(torch.nn.Module):"
+    InstanceAttrIdentifier ":" ["torch.jit.Final("] TSType [")"]
+    ...
+
+where ``InstanceAttrIdentifier`` is the name of an instance attribute and ``torch.jit.Final`` indicates that the attribute cannot be re-assigned outside
+of ``__init__`` or overridden in subclasses.
+
+**Example**
+
+In this example, ``a`` is not annotated and assumes the default type of ``TensorType``, ``b`` is annotated as type ``int``, and the return type is not
+annotated and is automatically inferred as type ``TensorType`` (based on the type of the value being returned).
+
+::
+
+    import torch
+
+    class MyModule(torch.nn.Module):
+        offset_: int
+
+    def __init__(self, offset):
+        self.offset_ = offset
+
+    ...
+
+
+
+Type Annotation APIs
+^^^^^^^^^^^^^^^^^^^^
+
+``torch.jit.annotate(T, expr)``
+"""""""""""""""""""""""""""""""
+This API annotates type ``T`` to an expression ``expr``. This is often used when the default type of an expression is not the type intended by the programmer.
+For instance, an empty list (dictionary) has the default type of ``List[TensorType]`` (``Dict[TensorType, TensorType]``) but sometimes it may be used to initialize
+a list of some other types. Another common use case is for annotating the return type of ``tensor.tolist()``.
+
+**Example**
+
+In this example, ``[]`` is declared as a list of integers via ``torch.jit.annotate`` (instead of assuming ``[]`` to be the default type of ``List[TensorType]``).
+
+::
+
+    import torch
+    from typing import List
+
+    def f(append: bool, val: int):
+        l = torch.jit.annotate(List[int], [])
+        if append:
+            l.append(val)
+            return l
+        else:
+            return None
+
+    m = torch.jit.script(f)
+    print("Eager:", f(True, 1), f(False, 1))
+    print("TorchScript:", m(True, 1), m(False, 1))
+
+TODO: Link to section about ``torch.jit.annotate``.
```

Comment:
Fix the TODO

Reply:
It cannot be done until `torch.jit.annotate` section lands.